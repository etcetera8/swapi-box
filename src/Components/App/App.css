.App {
  height: 100%;
  margin: 0;
  padding: 0;
}

.button-container {
  display: flex;
  width: 80%;
  flex-direction: row;
  justify-content: space-between;
  margin: 0 auto; 
}

#num {
  color: #AD1F1E;
}

.error-status {
  color: #AD1F1E;
}

.error-div {
  background: #000;
  height: 100vh;
}