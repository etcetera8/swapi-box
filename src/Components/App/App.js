import React, { Component } from 'react';
import ScrollContainer from '../ScrollContainer/ScrollContainer';
import CardContainer from '../CardContainer/CardContainer';
import Button from '../Button/Button';
import Cleaner from '../../helper';
import './App.css';

class App extends Component {
  constructor(props) {
    super(props)
    this.state = {
      scrollText: {},
      people: [],
      planets: [],
      vehicles: [],
      favorites: [],
      activeCategory: null
    }
    this.cleaner = new Cleaner()
  }

  componentDidMount() {
    this.cleaner = new Cleaner();
    //this.apiCall('films')
  }

  apiCall(resource) {
    const randNum = this.cleaner.randomMovieNumber();
    fetch(`https://swapi.co/api/${resource}/${randNum}`)
    .then( response => response.json())
    .then( json => {
      this.setState({ scrollText: this.cleaner.randomMovieCall(json) })
    })
  }

  setPeopleState = async () => {
    if (!localStorage.getItem('people')) {
      const people = await this.cleaner.getPeople()
      this.setState({people})
      localStorage.setItem('people', JSON.stringify(people))
    } else {
      this.setState({people: JSON.parse(localStorage.getItem("people"))})
    }
    this.setState({activeCategory: "people"})
  }

  setPlanetState = async () => {
    if (!localStorage.getItem('planets')) { 
      const planets = await this.cleaner.getPlanets()
      this.setState({planets})
      localStorage.setItem('planets', JSON.stringify(planets))
    } else {
      console.log("from loc");
      this.setState({planets: JSON.parse(localStorage.getItem('planets'))})
    }
    this.setState({activeCategory: "planets"})
  }

  setVehicleState = async () => {
    if (!localStorage.getItem('vehicles')) {
      const vehicles = await this.cleaner.getVehicles()
      this.setState({vehicles})
      localStorage.setItem('vehicles', JSON.stringify(vehicles))
    } else {
      this.setState({vehicles: JSON.parse(localStorage.getItem('vehicles'))})
    }

    this.setState({activeCategory: "vehicles"})
  }

  setFavoriteState = (cardName, category) => {
    const {favorites} = this.state;

    const target = this.state[category].find(card => { 
      card.name === cardName ? card.favorite = true : null;
      return card.name === cardName;
    });
    const favoriteCards = [...favorites, target];
    this.setState({favorites: favoriteCards });
  }

  setCategory = () => {
    this.setState({activeCategory: "favorites"})
  }

  activeStatus = (category) => {
    return this.state.activeCategory === category ?  
      `active${category}` : ""
  }

  render() {
    return (
      <div className="App">
        <ScrollContainer movieData={this.state.scrollText}/>
        <nav className="button-container">
          <Button
            name="people"
            resourceCall={this.setPeopleState}
            className={this.activeStatus('people')}
            />
          <Button 
            name="planets"
            resourceCall={this.setPlanetState}
            className={this.activeStatus('planets')}
            />
          <Button 
            name="vehicles"
            resourceCall={this.setVehicleState}
            className={this.activeStatus('vehicles')}/>
          <button 
            onClick={this.setCategory}
            className={`Button ${this.activeStatus('favorites')}`}
            >
            {this.state.favorites.length} Favorites
          </button>
        </nav>
        <CardContainer
          activeCategory={this.state.activeCategory}
          setFavorite={this.setFavoriteState}
          favorites={this.state.favorites}
          people={this.state.people}
          planets={this.state.planets}
          vehicles={this.state.vehicles} />
      </div>
    );
  }
}

export default App;
