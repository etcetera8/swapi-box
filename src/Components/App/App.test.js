/* eslint-disable */
import React from 'react';
import ReactDOM from 'react-dom';
import App from './App';
import {shallow } from 'enzyme';

window.fetch = jest.fn();

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<App />, div);
  ReactDOM.unmountComponentAtNode(div);
});

describe('App', ()=> {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<App />);
    global.localStorage = {
      getItem(keyword) {
        if (!global.localStorage[keyword]) {
          return null;
        }
        return JSON.stringify(global.localStorage[keyword]);
      },
      setItem(keyword, value) {
        global.localStorage[keyword] = value;
      }
    };
  });

  it('has initial state with a scrollText object, people, planets, vehicles and favorites array, and activeCategory default set to null', () => {
    const wrapper = shallow(<App />);
    const expectedState = { 
      scrollText: {},
      people: [],
      planets: [],
      vehicles: [],
      favorites: [],
      activeCategory: null,
      errorStatus: false };
    expect(wrapper.state()).toEqual(expectedState);
  });

  it('the apiCall function should setState of scroll text to an obj', async () => {
    const mock = {
      'title': "Attack of the Clones",
      'episode_id': 2, 
      'opening_crawl': "There is unrest in the Galactic", 
      'release_date': "02-07-2004", 
      'producer': "Rick McCallum"
    };

    window.fetch = jest.fn().mockImplementation( () => {
      Promise.resolve({
        status: 200,
        results: mock
      });
    });
    const wrapper = shallow(<App />);
    const inst = wrapper.instance();
  
    await inst.randomMovieScroll('films');
    expect(typeof wrapper.state().scrollText).toEqual('object');
    expect(Object.keys(wrapper.state().scrollText)).toEqual(['scrollText', 'title', 'releaseDate', 'episode']);
  });

  it("the setPeopleState function should update the people state to the array of people", async () => {
    const mockObj = {"name": 'luke', "homeworld": 'tattooine'};
    window.fetch = jest.fn().mockImplementation(() => 
      Promise.resolve({
        status: 200,
        json: () => Promise.resolve({
          status: 200,
          results: [mockObj]
        })
      }));
    const inst = wrapper.instance();
    await inst.setPeopleState();

    expect(wrapper.state().people.length).toBeGreaterThan(0);
  });

  it("the setPlanetState function should update the planet state to the array of planets", async () => {
    const mockArray = [{
      "name": 'tattooine', 
      "residents": ["place", "anotherPlace"], 
      "population": '10', 
      "climate": "hot", 
      "terrain": 'permafrost', 
      'notKey': 'data not needed'
    }];
    const expected = [{
      "category": "planets", 
      "climate": "hot", 
      "favorite": false, 
      "name": "tattooine", 
      "population": "10", 
      "residents": [undefined, undefined], 
      "terrain": "permafrost"
    }];
    window.fetch = jest.fn().mockImplementation(() => 
      Promise.resolve({
        status: 200,
        json: () => Promise.resolve({
          status: 200,
          results: mockArray
        })
      }));
    const inst = wrapper.instance();
    await inst.setPlanetState();

    expect(wrapper.state().planets.length).toBeGreaterThan(0);
    expect(wrapper.state().planets).toEqual(expected);
  });

  it("the setVehicleState function should update the vehicle state to array of vehicles", async () => {
    const mockObjectArray = [{
      name: 'sand buggy', 
      model:'beater', 
      vehicle_class: 'off-road', 
      passengers: '4', 
      notKey: 'should not be in final'
    }];
    const expectedArray = [{ 
      name: 'sand buggy', 
      vehicle_class: 'off-road', 
      passengers: '4', model: 'beater', 
      favorite: false, 
      category: 'vehicles'
    }];
    window.fetch = jest.fn().mockImplementation( () => 
      Promise.resolve({
        status: 200,
        json: () => 
          Promise.resolve({
            status: 200, 
            results: mockObjectArray
          })
      }));

    const inst = wrapper.instance();
    await inst.setVehicleState();
    expect(wrapper.state().vehicles.length).toBeGreaterThan(0);
    expect(wrapper.state().vehicles).toEqual(expectedArray);
  });

  it("the setFavoriteState should update the favoriteState when a card is clicked on and switch the key from false to favorite", () => {
    wrapper.setState({
      people: [{
        name:'luke', 
        species: 'human', 
        homeworld: 'naboo', 
        population: '10', 
        favorite: false
      }]
    });
    const expected = {
      name:'luke', 
      species: 'human', 
      homeworld: 'naboo', 
      population: '10', 
      favorite: true
    };
    const inst = wrapper.instance();
    
    expect(wrapper.state().people[0].favorite).toEqual(false);
    inst.setFavoriteState('luke', 'people');
    expect(wrapper.state().people[0].favorite).toEqual(true);

    expect(wrapper.state().favorites.length).toBeGreaterThan(0);
    expect(wrapper.state().favorites[0]).toEqual(expected);
    expect(wrapper.state().people[0].favorite).toEqual(true);
  });

  it("the setFavoriteState should take out the card if ran on it again", () => {
    const personArray = [{
      name:'luke', 
      species: 'human', 
      homeworld: 'naboo', 
      population: '10', 
      favorite: true
    }];
    wrapper.setState({favorites: personArray, people: personArray});
    const inst = wrapper.instance();

    expect(wrapper.state().favorites.length).toEqual(1);
    expect(wrapper.state().people).toEqual(personArray);

    inst.setFavoriteState('luke', 'people');

    expect(wrapper.state().people).toEqual(personArray);
    expect(wrapper.state().favorites.length).toEqual(0);
  });

  it("the setCategory should be null at first then update the state activeCategory to the argument passed it", () => {
    expect(wrapper.state().activeCategory).toEqual(null);
    const inst = wrapper.instance();
    inst.setCategory('planets');
    expect(wrapper.state().activeCategory).toEqual('planets');
  });

  it("the activeStatus function should return a blank string if the states activeCategory doesn't match the category passed in", () => {
    const inst = wrapper.instance();
    let className = inst.activeStatus('people');
    expect(className).toEqual('');
    wrapper.setState({activeCategory: 'people'});
    className = inst.activeStatus('people');
    expect(className).toEqual('activepeople');
  });

  it("should match the snapshot", () => {
    const personArray = [{
      name:'luke', 
      species: 'human', 
      homeworld: 'naboo', 
      population: '10', 
      favorite: true
    }];
    const scrollObject = {
      scrollText: "Once upon a time", 
      title: "benjamins buttons", 
      releaseDate:"05-07-87", 
      episode: "v"
    };
    wrapper.setState({
      scrollText: scrollObject,
      people: personArray,
      activeCategory: 'people'
    });
    expect(wrapper).toMatchSnapshot();
  });
});
