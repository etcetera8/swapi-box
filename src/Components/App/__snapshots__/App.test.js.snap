// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match the snapshot 1`] = `
<div
  className="App"
>
  <ScrollContainer
    movieData={
      Object {
        "episode": "v",
        "releaseDate": "05-07-87",
        "scrollText": "Once upon a time",
        "title": "benjamins buttons",
      }
    }
  />
  <nav
    className="button-container"
  >
    <Button
      className="activepeople"
      name="people"
      resourceCall={[Function]}
    />
    <Button
      className=""
      name="planets"
      resourceCall={[Function]}
    />
    <Button
      className=""
      name="vehicles"
      resourceCall={[Function]}
    />
    <button
      className="Button "
      onClick={[Function]}
    >
      <span
        id="num"
      >
        0
      </span>
       Favorites
    </button>
  </nav>
  <CardContainer
    activeCategory="people"
    favorites={Array []}
    people={
      Array [
        Object {
          "favorite": true,
          "homeworld": "naboo",
          "name": "luke",
          "population": "10",
          "species": "human",
        },
      ]
    }
    planets={Array []}
    setFavorite={[Function]}
    vehicles={Array []}
  />
</div>
`;
